### PR TITLE
Build updates

### PR DESCRIPTION
**Pull request by poelzi**
**How to categorize this PR?**

/area dev-productivity
/kind cleanup
/priority normal

**What this PR does / why we need it**:

Better portability to other OSs

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @poelzi for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: Makefile [hunk 1]
```diff
@@ -162,12 +167,17 @@ docker-push:
...
+	@true
+
+.PHONY: vendor-requirements
+vendor-requirements:
 	@go install -mod=vendor github.com/onsi/ginkgo/ginkgo
 	@go install -mod=vendor github.com/ahmetb/gen-crd-api-reference-docs
 	@go install -mod=vendor github.com/golang/mock/mockgen
 	@go install -mod=vendor sigs.k8s.io/controller-tools/cmd/controller-gen
 	@GO111MODULE=off go get github.com/prometheus/prometheus/cmd/promtool
+install-extra-requirements:
```
Comment (timuthy) [inline Makefile:164]:
```suggestion

.PHONY: install-extra-requirements
install-extra-requirements:
```

File: hack/get-build-ld-flags.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash -e
+#!/usr/bin/env -S bash -e
```
Comment (vpnachev) [inline hack/get-build-ld-flags.sh]: What is `-S` flag about?

```bash
$ /usr/bin/env -S bash -e
/usr/bin/env: invalid option -- 'S'
Try '/usr/bin/env --help' for more information.
```

```bash
$ /usr/bin/env --version
env (GNU coreutils) 8.28
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Richard Mlynarik and David MacKenzie.
```


```bash
$ docker run -it --rm alpine:3.12.3 /usr/bin/env -S
/usr/bin/env: unrecognized option: S
BusyBox v1.31.1 () multi-call binary.

Usage: env [-iu] [-] [name=value]... [PROG ARGS]

Print the current environment or run PROG after setting up
the specified environment

        -, -i   Start with an empty environment
        -u      Remove variable from the environment

```

**Comment by rfranzke** [comment]
@poelzi do you plan to continue on this PR?

File: hack/get-build-ld-flags.sh (hunk 2 above)
Comment (poelzi) [reply hack/get-build-ld-flags.sh]: That seems to be an option in newer env versions. I used a more compatible version now.

File: Makefile [hunk 3]
```diff
@@ -90,6 +92,17 @@ remote-garden-down:
 	# Remove docker containers and networks
 	@-./hack/local-development/remote-garden/cleanup $(REMOTE_GARDEN_LABEL)
 
+.PHONY: start-all-servers
+start-all-servers:
```
Comment (vpnachev) [inline Makefile]:
```suggestion
.PHONY: start-all
start-all
```

This target starts also some controllers and they are not acting as servers, but just clients.

File: Makefile [hunk 4]
```diff
@@ -90,6 +92,17 @@ remote-garden-down:
...
+	# start a screen session and open all servers in named tabs
+	echo "starting screen session with all servers"
+	@screen -ls gardener-dev || screen -AdmS gardner-dev -t tab0 bash
+	@screen -S gardener-dev -X screen -t apiserver bash -c "make --debug=j start-apiserver; exec bash"
+	@screen -S gardener-dev -X screen -t controller bash -c "make --debug=j start-controller-manager; exec bash"
+	@screen -S gardener-dev -X screen -t scheduler bash -c "make --debug=j start-scheduler; exec bash"
+	@screen -S gardener-dev -X screen -t gardenlet bash -c "make --debug=j start-gardenlet; exec bash"
```
Comment (vpnachev) [inline Makefile]: Please, add `admission-controller` and `seed-admission-controller`.

File: Makefile [hunk 5]
```diff
@@ -162,12 +175,19 @@ docker-push:
 #####################################################################
 
 .PHONY: install-requirements
-install-requirements:
+install-requirements: vendor-requirements install-extra-requirements
+	@true
```
Comment (vpnachev) [inline Makefile]: Why is this `@true` necessary?
The `verify` target is also just a wrapper to other dependencies and it does not need it.

File: hack/usage/retry-shoot-operation [hunk 6]
```diff
@@ -13,5 +13,5 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+set -eu
```
Comment (vpnachev) [inline hack/usage/retry-shoot-operation:16]:
```suggestion
set -e
set -u
```
However, I have no string opinion about this. So whatever you decide.

File: Makefile [hunk 7]
```diff
@@ -90,6 +92,17 @@ remote-garden-down:
 	# Remove docker containers and networks
 	@-./hack/local-development/remote-garden/cleanup $(REMOTE_GARDEN_LABEL)
 
+.PHONY: start-all-servers
+start-all-servers:
+	# start a screen session and open all servers in named tabs
+	echo "starting screen session with all servers"
+	@screen -ls gardener-dev || screen -AdmS gardner-dev -t tab0 bash
```
Comment (vpnachev) [inline Makefile]: Add several lines about how to install `screen` [here ](https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md).

**Review by vpnachev — CHANGES_REQUESTED** [review]
I cannot make `screen` to work
```bash
$ make start-all-servers
# start a screen session and open all servers in named tabs
echo "starting screen session with all servers"
starting screen session with all servers
No Sockets found in /run/screen/S-vpnachev.

No screen session found.
Makefile:97: recipe for target 'start-all-servers' failed
make: *** [start-all-servers] Error 1

$ screen -v
Screen version 4.06.02 (GNU) 23-Oct-17
```

I am running on up-to-date Ubuntu 18.04 and screen was installed via the apt package manager. Could you advise how to get it working?

File: Makefile [hunk 8]
```diff
@@ -58,6 +59,19 @@ remote-garden-up:
...
+start-all:
+	# start a screen session and open all servers in named tabs
+	echo "starting screen session with all servers"
+	@screen -ls $(DEV_SCREEN_NAME) || screen -AdmS $(DEV_SCREEN_NAME) -t tab0 bash
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t apiserver bash -c "make --debug=j start-apiserver; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t controller bash -c "make --debug=j start-controller-manager; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t scheduler bash -c "make --debug=j start-scheduler; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t gardenlet bash -c "make --debug=j start-gardenlet; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t admission bash -c "make --debug=j admission-controller; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t seed bash -c "make --debug=j seed-admission-controller; exec bash"
```
Comment (vpnachev) [inline Makefile]:
```suggestion
	@screen -S $(DEV_SCREEN_NAME) -X screen -t admission bash -c "make --debug=j start-admission-controller; exec bash"
	@screen -S $(DEV_SCREEN_NAME) -X screen -t seed-admission bash -c "make --debug=j start-seed-admission-controller; exec bash"
```

File: Makefile [hunk 9]
```diff
@@ -58,6 +59,19 @@ remote-garden-up:
 remote-garden-down:
 	@./hack/local-development/remote-garden/stop.sh $(REMOTE_GARDEN_LABEL)
 
+.PHONY: start-all
+start-all:
+	# start a screen session and open all servers in named tabs
+	echo "starting screen session with all servers"
+	@screen -ls $(DEV_SCREEN_NAME) || screen -AdmS $(DEV_SCREEN_NAME) -t tab0 bash
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t apiserver bash -c "make --debug=j start-apiserver; exec bash"
```
Comment (vpnachev) [inline Makefile:68]: It would be nice to wait the API server to start and to be reachable, e.g. `kubectl get shoot` to not return errors, before to start the other components.

**Review by vpnachev — DISMISSED** [review]
Thanks!
I have added my comments in-line.

File: Makefile [hunk 10]
```diff
@@ -146,6 +166,7 @@ revendor:
 	@GO111MODULE=on cd third_party/kube-scheduler/v19; go mod tidy
 	@GO111MODULE=on cd third_party/kube-scheduler/v20; go mod tidy
 	@GO111MODULE=on cd third_party/kube-scheduler; go mod tidy
+	@cd vendor/k8s.io/code-generator; go mod vendor
```
Comment (rfranzke) [inline Makefile:172]: Hm, why is this needed?

Comment (poelzi) [reply Makefile:172]: On my NixOS I could not get it building because when the code-generator was build, it never had it's dependencies available and was failing with typical cryptic go error messages. It complained that it needed vendoring and this fixed it.
go does not walk, limps ....

File: Makefile (hunk 9 above)
Comment (poelzi) [reply Makefile:68]: done

File: Makefile (hunk 10 above)
Comment (vpnachev) [reply Makefile:172]: What version of go are you using?
Is the gardener in your ${GOPATH}?

File: docs/development/local_setup.md [hunk 11]
```diff
@@ -127,6 +127,36 @@ The Gardener repository and all the above-mentioned tools (git, golang, kubectl,
...
+
+```bash
+nix-shell
+```
+
+You can use this environment from a IDE with commands like:
+
+```bash
+nix-shell --command "make generate"
+```
```
Comment (vpnachev) [inline docs/development/local_setup.md]: Could you point the advantages of this distro/shell?

File: hack/nix/spiff/deps.nix [hunk 12]
```diff
@@ -0,0 +1,246 @@
+# file generated from Gopkg.lock using dep2nix (https://github.com/nixcloud/dep2nix)
```
Comment (vpnachev) [inline hack/nix/spiff/deps.nix:1]: I guess this file needs to be updated basically every time when go.mod is updated, am I right?

File: shell.nix [hunk 13]
```diff
@@ -0,0 +1,48 @@
+{ pkgs ? import <nixpkgs> { } }:
+let
+  inherit (pkgs) lib buildGoPackage fetchFromGitHub;
+
+  spiff = buildGoPackage rec {
+    pname = "minio-exporter";
+    version = "1.6.0-beta-1";
+    rev = "v${version}";
+
+    goPackagePath = "github.com/mandelsoft/spiff";
```
Comment (vpnachev) [inline shell.nix:10]: interesting...

File: docs/development/local_setup.md (hunk 11 above)
Comment (poelzi) [reply docs/development/local_setup.md]: It is fully reproducible and package compilation and configuration is done by the same system. A user package configuration can also trigger that a package is compiled with different flags if necessary without any user interaction. You don't need NixOS to enjoy the power of nix, it can be installed on other operating systems.

https://nixos.org/guides/how-nix-works.html

File: hack/nix/spiff/deps.nix (hunk 12 above)
Comment (poelzi) [reply hack/nix/spiff/deps.nix:1]: If spiff is updated to the newer go module support we can remove this file and use buildGoModule instead which uses a single hash value over all dependencies.
In nix, everything is about reproducibility. You can have multiple versions of any library/program installed in parallel, which helps a lot when testing different environments.

File: Makefile (hunk 10 above)
Comment (poelzi) [reply Makefile:172]: go version go1.15.5 linux/amd64
And yes, I'm unfortunately in ~/go/github.com/gardener/gardener

Comment (rfranzke) [reply Makefile:172]: Shouldn't it be `~/go/src/...`? what's the value of your `$GOPATH`?

File: shell.nix [hunk 14]
```diff
@@ -0,0 +1,48 @@
+{ pkgs ? import <nixpkgs> { } }:
```
Comment (rfranzke) [inline shell.nix:1]: Why is this in the root? Can it be moved elsewhere, to a dedicated folder only containing the "nix" stuff? `hack/nix`?

Comment (poelzi) [reply shell.nix:1]: Because shell-nix searches for default.nix and shell.nix in the current directory. That is the default location for all projects so far I have seen that ship with a nix file.

File: docs/development/local_setup.md (hunk 11 above)
Comment (poelzi) [reply docs/development/local_setup.md]: I linked in the docs to the nix page

Comment (vpnachev) [reply docs/development/local_setup.md]: Well, we have https://github.com/gardener/gardener/issues/2143 exactly for this purpose.

File: shell.nix (hunk 13 above)
Comment (vpnachev) [reply shell.nix:10]: /invite @mandelsoft 

as he seems to have experience with the changes in this PR.

File: Makefile (hunk 10 above)
Comment (poelzi) [reply Makefile:172]: @rfranzke ups, it is of course ~/go/src/github.com/gardener/gardener

**Comment by poelzi** [comment]
I plan on extending the shell.nix file with time to have more useful feature as I get more accounted with everything.
For example, this [shell.nix](https://github.com/poelzi/mixxx/blob/nix-updates-2.3/shell.nix) I maintain for a c++ project. It allows for example to build using different compiler versions (gcc, clang, clazy), different dependencies with optional debug builds etc.
It is super comfy to use and a new developer has to do nothing to get everything working. You don't even need no be root.

File: docs/development/local_setup.md (hunk 11 above)
Comment (poelzi) [reply docs/development/local_setup.md]: @vpnachev docker containers are hardly reproduce-able, you can only store your build images indefinitely and hope you never loose them. If there is any apt-update whatsoever, any curl download without check-sums, your build is just similar, not identical.

File: Makefile [hunk 15]
```diff
@@ -59,6 +60,24 @@ remote-garden-up:
...
+	# start a screen session and open all servers in named tabs
+	echo "starting screen session with all servers"
+	@screen -ls $(DEV_SCREEN_NAME) || screen -AdmS $(DEV_SCREEN_NAME) -t tab0 bash
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t apiserver bash -c "make --debug=j start-apiserver; exec bash"
+	@./hack/wait-for-apiserver.sh
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t controller bash -c "make --debug=j start-controller-manager; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t scheduler bash -c "make --debug=j start-scheduler; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t gardenlet bash -c "make --debug=j start-gardenlet; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t admission bash -c "make --debug=j start-admission-controller; exec bash"
+	@screen -S $(DEV_SCREEN_NAME) -X screen -t seed-admission bash -c "make --debug=j start-seed-admission-controller; exec bash"
```
Comment (rfranzke) [inline Makefile:74]: I guess this line should be removed as we would not start the seed-admission-controller (GSAC) locally (usually, at least not in conjunction with the others) because it's being deployed by the gardenlet and runs in the seed as a container.
For local development, the GSAC would be started locally, of course, but I guess it's not necessary in this "start-all"/screen session here.

File: Makefile [hunk 16]
```diff
@@ -131,19 +150,26 @@ docker-push:
 #####################################################################
 
 .PHONY: install-requirements
-install-requirements:
+install-requirements: vendor-requirements install-extra-requirements
+
+.PHONY: vendor-requirements
+vendor-requirements:
```
Comment (rfranzke) [inline Makefile:156]:
```suggestion
install-requirements: install-go-requirements install-extra-requirements

.PHONY: install-go-requirements
install-go-requirements:
```

File: Makefile [hunk 17]
```diff
@@ -131,19 +150,26 @@ docker-push:
...
+.PHONY: install-extra-requirements
+install-extra-requirements:
 	@./hack/install-requirements.sh
 
 .PHONY: revendor
 revendor:
 	@GO111MODULE=on go mod vendor
 	@GO111MODULE=on go mod tidy
 	@curl -sSLo hack/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/$(CR_VERSION)/hack/setup-envtest.sh
+	@cd vendor/k8s.io/code-generator; go mod vendor
```
Comment (rfranzke) [inline Makefile:172]: /invite @ialidzhikov - do you have opinions on this change?

File: docs/development/local_setup.md [hunk 18]
```diff
@@ -127,6 +127,36 @@ The Gardener repository and all the above-mentioned tools (git, golang, kubectl,
...
+
+To install screen on *Debian/Ubuntu*
+
+```bash
+apt install screen
+```
+
+## NixOS/Nix support
+
+Gardner contains a `shell.nix` file for development purposes. The nix package manager
```
Comment (rfranzke) [inline docs/development/local_setup.md:147]:
```suggestion
Gardener contains a `shell.nix` file for development purposes. The nix package manager
```

File: Makefile (hunk 17 above)
Comment (ialidzhikov) [reply Makefile:172]: The same question is already asked above but I still don't get why this is required. We are using `./vendor/k8s.io/code-generator/generate-groups.sh` and `./vendor/k8s.io/code-generator/generate-internal-groups.sh` scripts and I don't see why such change is required.
@poelzi , if you believe this is something specific to your env, can you try to reproduce with https://github.com/kubernetes/sample-apiserver locally and open issue to https://github.com/kubernetes/kubernetes?

**Review by ialidzhikov — COMMENTED** [review]
@poelzi , can you outline what this PR does? It makes local dev setup compatible for `nix-shell`? If yes, why we should be interested in something similar instead of https://github.com/gardener/gardener/issues/2143?

**Comment by rfranzke** [comment]
Are you still planning to work on this PR?
Generally, I have my doubts about the maintenance of these files. I doubt that somebody of @gardener/gardener-maintainers will switch to it, so not sure how sustainable all of this is.

**Comment by rfranzke** [comment]
OK, let's
/close
the PR for now. I don't see much progress anymore, and while we highly appreciate your contribution, still I guess most maintainers have the same doubts as outlined in https://github.com/gardener/gardener/pull/3397#issuecomment-775681907. Please feel free to /reopen it if you see necessity. I'm not against it in general because I think most changes are actually useful, but probably the "nix" related files would just become outdated over time, so maybe we can omit them in this case?
